modules/gapi/src/compiler/passes/perform_substitution.cpp

```diff
@@ -0,0 +1,72 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+#include "pattern_matching.hpp"
+
+namespace cv { namespace gimpl {
+namespace {
+using Graph = GModel::Graph;
+
+template<typename It, typename Callable>
+void erase(Graph& g, It first, It last, Callable get) {
+    for (; first != last; ++first) {
+        ade::NodeHandle node = get(first);
+        if (node == nullptr) continue;  // some nodes might already be erased
+        g.erase(node);
+    }
+}
+}  // anonymous namespace
+
+void performSubstitution(Graph& graph,
+    const SubgraphMatch& patternToGraph, const SubgraphMatch& patternToSubstitute) {
+    // substitute input nodes
+    for (const auto& inputNodePair : patternToGraph.inputDataNodes) {
+        // Note: we don't replace input DATA nodes here, only redirect their output edges
+        const auto& patternDataNode = inputNodePair.first;
+        const auto& graphDataNode = inputNodePair.second;
+        const auto& substituteDataNode = patternToSubstitute.inputDataNodes.at(patternDataNode);
+        GModel::redirectReaders(graph, substituteDataNode, graphDataNode);
+    }
+
+    // substitute output nodes
+    for (const auto& outputNodePair : patternToGraph.outputDataNodes) {
+        // Note: we don't replace output DATA nodes here, only redirect their input edges
+        const auto& patternDataNode = outputNodePair.first;
+        const auto& graphDataNode = outputNodePair.second;
+        const auto& substituteDataNode = patternToSubstitute.outputDataNodes.at(patternDataNode);
+        for (auto e : graphDataNode->inEdges()) {
+            graph.erase(e);
+        }
+        GModel::redirectWriter(graph, substituteDataNode, graphDataNode);
+    }
```

Comment:
writers: graphDataNode here is OUTPUT for pattern/substitute and so we need to connect "last level" OP nodes from substitute to it (without substituting the DATA node itself - otherwise, we get a new output that doesn't match to user's output data)

---
It's fused (2nd case): this way it is much simpler to implement both substitute nodes construction && replacement of pattern with substitute.

First case (I have it in commit history as well): we'd just need to traverse the substitute graph and re-create each node in main graph + connect them as in substitute as we go through